File: TinkerRocketApp/TinkerRocketApp/Maps/Offline/TileMath.swift
//
//  TileMath.swift
//  TinkerRocketApp
//
//  Offline maps — Trial 2 (see docs/plans/offline-maps.md).
//
//  Slippy-map tile math: convert a region (center + radius + zoom range) into
//  the set of (z,x,y) tiles to download, and count them for the pre-download
//  size estimate.
//

import Foundation
import CoreLocation

/// A single web-mercator tile coordinate.
struct TileXYZ: Hashable {
    let z: Int
    let x: Int
    let y: Int
}

/// What to download: a circle (center + radius) over a span of zoom levels.
struct RegionSpec {
    var center: CLLocationCoordinate2D
    var radiusMeters: Double
    var minZoom: Int = 10
    var maxZoom: Int

    /// Lat/lon bounding box of the circle (small-angle approximation — fine for
    /// the few-km radii used here).
    var bbox: (latMin: Double, latMax: Double, lonMin: Double, lonMax: Double) {
        let dLat = radiusMeters / 111_320.0
        let cosLat = max(0.01, cos(center.latitude * .pi / 180))
        let dLon = radiusMeters / (111_320.0 * cosLat)
        return (center.latitude - dLat, center.latitude + dLat,
                center.longitude - dLon, center.longitude + dLon)
    }
}

enum TileMath {
    static func tileX(lon: Double, z: Int) -> Int {
        let n = Double(1 << z)
        return Int(floor((lon + 180.0) / 360.0 * n))
    }

    static func tileY(lat: Double, z: Int) -> Int {
        let n = Double(1 << z)
        let r = lat * .pi / 180.0
        return Int(floor((1.0 - asinh(tan(r)) / .pi) / 2.0 * n))
    }

    private static func clamp(_ v: Int, _ lo: Int, _ hi: Int) -> Int {
        min(max(v, lo), hi)
    }

    /// Inclusive tile rectangle for a bbox at one zoom level.
    private static func rect(_ b: (latMin: Double, latMax: Double, lonMin: Double, lonMax: Double),
                             z: Int) -> (xMin: Int, xMax: Int, yMin: Int, yMax: Int) {
        let n = 1 << z
        // North (latMax) maps to the smaller y; south (latMin) to the larger y.
        return (clamp(tileX(lon: b.lonMin, z: z), 0, n - 1),
                clamp(tileX(lon: b.lonMax, z: z), 0, n - 1),
                clamp(tileY(lat: b.latMax, z: z), 0, n - 1),
                clamp(tileY(lat: b.latMin, z: z), 0, n - 1))
    }

    /// Total tile count for the region (for the size estimate; no allocation).
    static func tileCount(for region: RegionSpec) -> Int {
        let b = region.bbox
        var count = 0
        for z in min(region.minZoom, region.maxZoom)...region.maxZoom {
            let r = rect(b, z: z)
            count += (r.xMax - r.xMin + 1) * (r.yMax - r.yMin + 1)
        }
        return count
    }

    /// The full list of tiles to download for the region.
    static func tiles(for region: RegionSpec) -> [TileXYZ] {
        let b = region.bbox
        var out: [TileXYZ] = []
        for z in min(region.minZoom, region.maxZoom)...region.maxZoom {
            let r = rect(b, z: z)
            for x in r.xMin...r.xMax {
                for y in r.yMin...r.yMax {
                    out.append(TileXYZ(z: z, x: x, y: y))
                }
            }
        }
        return out
    }
}

File: TinkerRocketAndroid/core/maps/src/main/kotlin/com/tinkerbug/tinkerrocket/maps/TileMath.kt
package com.tinkerbug.tinkerrocket.maps

import kotlin.math.asinh
import kotlin.math.cos
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.min
import kotlin.math.tan

/** A single web-mercator tile coordinate. */
public data class TileXYZ(val z: Int, val x: Int, val y: Int)

/**
 * What to download: a circle (center + radius) over a span of zoom levels —
 * port of iOS TileMath.swift RegionSpec.
 */
public data class RegionSpec(
    val centerLat: Double,
    val centerLon: Double,
    val radiusMeters: Double,
    val minZoom: Int = 10,
    val maxZoom: Int,
) {
    /**
     * Lat/lon bounding box of the circle (small-angle approximation — fine
     * for the few-km radii used here).
     */
    public val bbox: Bbox
        get() {
            val dLat = radiusMeters / 111_320.0
            val cosLat = max(0.01, cos(Math.toRadians(centerLat)))
            val dLon = radiusMeters / (111_320.0 * cosLat)
            return Bbox(centerLat - dLat, centerLat + dLat, centerLon - dLon, centerLon + dLon)
        }
}

public data class Bbox(val latMin: Double, val latMax: Double, val lonMin: Double, val lonMax: Double)

/** Slippy-map tile math — port of iOS TileMath.swift, formulas verbatim. */
public object TileMath {

    public fun tileX(lon: Double, z: Int): Int {
        val n = (1 shl z).toDouble()
        return floor((lon + 180.0) / 360.0 * n).toInt()
    }

    public fun tileY(lat: Double, z: Int): Int {
        val n = (1 shl z).toDouble()
        val r = Math.toRadians(lat)
        return floor((1.0 - asinh(tan(r)) / Math.PI) / 2.0 * n).toInt()
    }

    private fun clamp(v: Int, lo: Int, hi: Int) = min(max(v, lo), hi)

    /** Inclusive tile rectangle for a bbox at one zoom level. */
    private fun rect(b: Bbox, z: Int): IntArray {
        val n = 1 shl z
        // North (latMax) maps to the smaller y; south (latMin) to the larger y.
        return intArrayOf(
            clamp(tileX(b.lonMin, z), 0, n - 1),
            clamp(tileX(b.lonMax, z), 0, n - 1),
            clamp(tileY(b.latMax, z), 0, n - 1),
            clamp(tileY(b.latMin, z), 0, n - 1),
        )
    }

    /** Total tile count for the region (for the size estimate; no allocation). */
    public fun tileCount(region: RegionSpec): Int {
        val b = region.bbox
        var count = 0
        for (z in min(region.minZoom, region.maxZoom)..region.maxZoom) {
            val (xMin, xMax, yMin, yMax) = rect(b, z)
            count += (xMax - xMin + 1) * (yMax - yMin + 1)
        }
        return count
    }

    /** The full list of tiles to download for the region. */
    public fun tiles(region: RegionSpec): List<TileXYZ> {
        val b = region.bbox
        val out = mutableListOf<TileXYZ>()
        for (z in min(region.minZoom, region.maxZoom)..region.maxZoom) {
            val (xMin, xMax, yMin, yMax) = rect(b, z)
            for (x in xMin..xMax) {
                for (y in yMin..yMax) {
                    out += TileXYZ(z, x, y)
                }
            }
        }
        return out
    }

    private operator fun IntArray.component1() = this[0]
    private operator fun IntArray.component2() = this[1]
    private operator fun IntArray.component3() = this[2]
    private operator fun IntArray.component4() = this[3]
}
